documentation/src/docs/wire-protocol.md

```diff
@@ -9,104 +9,105 @@ You may obtain a copy of the License at
-2.  Length - Unsigned integer 4 bytes (Messages should be &lt; 2^24, but the upper bits remain zero). How many bytes of data from this point forward are part of this message. (Possibly zero, indicating there is no data)
+All the requests and replies have 8 byte header with two fields. (All data is written in **BigEndian** format).
 
-The remainder of the fields are specific to the type of message. A few important messages are listed below.
+- **Message Type**: A 4 byte Integer identifies the message type and determines the subsequent fields. (Note that the protocol can be extended by adding new types.)
+- **Length**:  A 4 byte Unsigned Integer (Messages should be less than 2^<sup>24</sup>, but the upper bits remain zero). The remaining bits followed by the upper bit are part of the message. (Possibly zero, indicating there is no data).
+The remainder of the fields are specific to the type of the message. A few important messages are listed below.
 
-General
--------
+# General
 
-### Partial Message - Request/Reply
+## KeepAlive - Request/Reply
 
-1.  Begin/Middle/End - Enum (1 byte)
-2.  Data
-A partial message is one that was broken up when being sent over the wire. (For any reason). The whole message is reconstructed by reading the partial messages in sequence and assembling them into a whole. It is not valid to attempt to start a new partial message before completing the previous one.
+- **Data**: Uninterpreted data of the length of the message. (Usually 0 byte)
 
-### KeepAlive - Request/Reply
+# Reading
 
-1.  Data - uninterpreted data of the length of the message. (Usually 0 bytes)
 
-Reading
--------
+## Read Segment - Request
 
-### Read Segment - Request
+1. Segment to read: A length of 2 byte String length, followed by that many bytes of Java's Modified UTF-8.
+2. Offset to read from:  8 byte Long.
+3. Suggested Length of Reply: 4 byte Integer. The clients can request for the required length to server (but the server may allot a different number of bytes).
+4. `delegationToken`: This was added to perform _auth_. It is an opaque-to-the-client token provided by the Controller that says it's allowed to make this call.
 
-1.  Segment to read - String (2 byte length, followed by that many bytes of Java's Modified UTF-8)
-2.  Offset to read from - Long (8 bytes)
-3.  Suggested Length of Reply - int (4 bytes)
-    1.  This is how much data the client wants. They won't necessarily get that much.
 
-### Segment Read - Reply
+## Segment Read - Reply
 
-1.  Segment that was read - String (2 byte length, followed by that many bytes of Java's Modified UTF-8)
-2.  Offset that was read from - Long (8 bytes)
-3.  Is at Tail - Boolean (1 bit)
-4.  Is at EndOfSegment - (1 bit)
-5.  Data - Binary (remaining length in message)
+1.  Segment that was read: A length of 2 byte String, followed by that many bytes of Java's Modified UTF-8).
```

Comment:
The original description was way more clear. This is worse.

Reply:
Yes! modified in all places.